lib/db/schemaupdater.go

```diff
@@ -629,5 +669,304 @@ func (db *schemaUpdater) updateSchemaTo12(_ int) error {
 		return err
 	}
 
+	if err := db.rewriteGlobals(t); err != nil {
+		return err
+	}
+
 	return t.Commit()
 }
+
+func (db *schemaUpdater) rewriteGlobals(t readWriteTransaction) error {
+	it, err := t.NewPrefixIterator([]byte{KeyTypeGlobal})
+	if err != nil {
+		return err
+	}
+	for it.Next() {
+		var vl VersionListDeprecated
+		if err := vl.Unmarshal(it.Value()); err != nil {
+			// If we crashed during an earlier migration, some version
+			// lists might already be in the new format: Skip those.
+			var nvl VersionList
+			if nerr := nvl.Unmarshal(it.Value()); nerr == nil {
+				continue
+			}
+			return err
+		}
+		if len(vl.Versions) == 0 {
+			if err := t.Delete(it.Key()); err != nil {
+				return err
+			}
+		}
+
+		var newVl VersionList
+		newPos := -1
+		oldPos := 0
+		var lastVersion protocol.Vector
+		for _, fv := range vl.Versions {
+			if fv.Invalid {
+				break
+			}
+			oldPos++
+			if lastVersion.Equal(fv.Version) {
+				newVl.RawVersions[newPos].Devices = append(newVl.RawVersions[newPos].Devices, fv.Device)
+				continue
+			}
+			newVl.RawVersions = append(newVl.RawVersions, newFileVersion(fv.Device, fv.Version, false, fv.Deleted))
+			lastVersion = fv.Version
+			newPos++
+		}
+		if oldPos == len(vl.Versions) {
+			if err := t.Put(it.Key(), mustMarshal(&newVl)); err != nil {
+				return err
+			}
+			continue
+		}
+		if len(newVl.RawVersions) == 0 {
+			fv := vl.Versions[oldPos]
+			newVl.RawVersions = []FileVersion{newFileVersion(fv.Device, fv.Version, true, fv.Deleted)}
+		}
+		newPos = 0
+	outer:
+		for _, fv := range vl.Versions[oldPos:] {
```

Comment:
I don't really understand what happens here or why. What's the reason we can't just trust the order the versions are currently in? Are the ordering requirements in the new format different than in the old?

Reply:
In the old format it's like two sorted lists appended to each other, first consisting of valid ones, second of invalid ones (e.g. something like `v:1, v:2, v:4, inv:1, inv:3, inv:4`). We handled the valid half above, now it's the invalid half. For which we need to check if a version already exists from the valid ones -> add it there, or if it's a new one, we need to get the correct position (e.g. in the given example, the position of `inv:3` isn't clear).

Reply:
But do we really need the fileinfo to determine the ordering? The conflict winner wasn't completely known with the previous order either, so could we not let that still be the case?

That would mean the migration is a simple aggregation on version (keep an index along the way) followed by a sort of the versions on version vector only

Reply:
We did (and still do) determine the conflict winner on insert to the global list, i.e. when versions are one of the two concurent values, get the files and call `WinsConflict` on them. Otherwise the puller would get the wrong file.

Reply:
I'm feeling dense here. Given we did/do that, the entries should already be in the right order in the old version list? The part I'm not getting is why we need more information than the old version list in order to determine the order in the new version list.

Reply:
Yeah right, I densely didn't consider that if `b.IsInvalid()` then `protocol.WinsConflict(a, b) == true`, and we do know whether a and b are (in-)valid - will simplify.  
If of interest: The perceived, incorrect problem was that the old version list consisted of two lists ordered by version (one for valid, one for invalid devices), while the new one is a single ordered list. With an example: Lets say the old list is `[{valid, A:2}, {valid, A:1}, {invalid, B:1}] `, now where does the third version go? It's not clear from the ordering itself, as it's a conflict, but it is considering it is invalid...